Add Zarr support

keller-mark opened the pull request:
Fixes #91 

These changes are from both me and @Artur-man 

The main public-facing changes here are:
- The `ZarrAnnData` class
- `read_zarr` and `write_zarr` top-level functions
- Support for `from_Seurat(output_class="ZarrAnnData")`
- Support for `from_SingleCellExperiment(output_class="ZarrAnnData")`

Internally:
- `read_zarr_helpers.R` is the zarr analog of `read_h5ad_helpers.R`
- `write_zarr_helpers.R` is the zarr analog of `write_h5ad_helpers.R`
- Test fixtures within `inst/extdata/example.zarr` (this makes the diff noisy, apologies)
- Lots of tests:
  - `test-Zarr-read.R` (35 new tests)
  - `test-Zarr-write.R` (70)
  - `test-ZarrAnnData.R` (26)
  - `test-h5ad-zarr.R` (17)
 
A number of these functions generate warnings in the R console that are intended to be followed up on to improve the code (and should probably be resolved as end users may not appreciate them), but the tests still pass despite these warnings.

Known things that are not implemented here:
- support for `recarrays`
- usage of `mode = c("r", "r+", "a", "w", "w-", "x")` parameter value

keller-mark committed:
WIP

keller-mark committed:
More tests passing

keller-mark committed:
Fix df read bug

keller-mark committed:
More tests passing after fixing zero-dimensional get bug in pizzarr

keller-mark committed:
WIP: writing

keller-mark committed:
Fix more tests

keller-mark committed:
Zarr df writing

keller-mark committed:
WIP: ZarrAnnData class

keller-mark committed:
Tests passing

keller-mark committed:
Tests that compare h5ad to zarr

keller-mark committed:
Use Rarr to read full numeric arrays

keller-mark committed:
Fix bugs. Add test for from_SingleCellExperiment with Zarr

keller-mark committed:
Add a to_dense param to ZarrAnnData constructor. Add overwrite params internally.

keller-mark committed:
Update

keller-mark committed:
Backwards dense/sparse

Artur-man committed:
Merge branch 'keller-mark/zarr' of https://github.com/keller-mark/anndataR into keller-mark/zarr

Artur-man committed:
Merge branch 'keller-mark/zarr' into zarr

Artur-man committed:
Simplify how obs and var names handled in ZarrAnnData (similar to #171)

Artur-man committed:
update extdata and documentation

Artur-man committed:
fix set/get zarr _index, update text example.zarr and update tests similar to HDF5AnnData

keller-mark committed:
Merge pull request #5 from Artur-man/zarr

Simplify how obs and var names handled in ZarrAnnData

keller-mark committed:
Merge

keller-mark committed:
Fix test

keller-mark committed:
Revert unnecessary changes

keller-mark committed:
Formatting

keller-mark committed:
Merge pull request #6 from keller-mark/keller-mark/zarr-2

More zarr-related changes

keller-mark committed:
Add comments

keller-mark committed:
Merge pull request #7 from keller-mark/keller-mark/comments

Update comments

rcannood commented on R/read_zarr.R:
```diff
@@ -0,0 +1,47 @@
+#' if (requireNamespace("SeuratObject", quietly = TRUE)) {
+#'   seurat <- read_zarr(store, to = "Seurat")
+#' }
+read_zarr <- function(
+    path,
+    to = c("InMemoryAnnData", "ZarrAnnData", "SingleCellExperiment", "Seurat"),
+    ...) {
+  to <- match.arg(to)
+
+  adata <- ZarrAnnData$new(path)
+
+  fun <- switch(to,
+    "SingleCellExperiment" = to_SingleCellExperiment,
+    "Seurat" = to_Seurat,
+    "InMemoryAnnData" = to_InMemoryAnnData,
+    "ZarrAnnData" = NULL
+  )
```
Should add HDF5AnnData as an option here as well. Which means that we should probably store this information elsewhere to avoid code duplication

```suggestion
read_zarr <- function(
    path,
    to = c("InMemoryAnnData", "HDF5AnnData", "SingleCellExperiment", "Seurat", "ZarrAnnData"),
    ...) {
  to <- match.arg(to)

  adata <- ZarrAnnData$new(path)

  fun <- switch(to,
    "SingleCellExperiment" = to_SingleCellExperiment,
    "Seurat" = to_Seurat,
    "InMemoryAnnData" = to_InMemoryAnnData,
    "HDF5AnnData" = to_HDF5AnnData,
    "ZarrAnnData" = NULL
  )

Artur-man replied:
done ... mixed this with the latest commit! https://github.com/scverse/anndataR/pull/190/commits/d192e68ed312f212476a5490c973619f08e7c8de

rcannood commented on R/read_zarr_helpers.R:
```diff
@@ -0,0 +1,510 @@
+  attrs <- g$get_attrs()$to_list()
+
+  if (!all(c("encoding-type", "encoding-version") %in% names(attrs))) {
+    path <- name
```
where are a lot of linting issues in this file -- could you run `lintr::lint_package()` and fix any issues that pop up?

Artur-man replied:
working on it now ...

Artur-man replied:
done ... did a full `lint_package` check and corrected some R check issues too!

rcannood commented on inst/extdata/example2.zarr/.zgroup:
This file should probably be removed

Artur-man replied:
done

rcannood reviewed (changes requested):
Fantastic work @keller-mark and @Artur-man !

I went through the PR for a first time and left some minor comments. I will review the code by running it a couple of times next :)

Artur-man committed:
remove unnecessary example zarr store

Artur-man commented:
@rcannood @keller-mark should we add `pizzarr` only under `Suggests` in `DESCRIPTION` and no need to add a `Remotes` list here ? 

```
Suggests:
    ...
    ...
    pizzarr
Remotes:
    keller-mark/pizzarr
```

Artur-man committed:
lintr and R check for zarr related utilities and functions, updated some documentation

keller-mark commented:
Good point, at minimum we should add a note to the Github README here https://github.com/scverse/anndataR/tree/main?tab=readme-ov-file#installation

Artur-man commented:
Ah, this is already added here, nevermind then only `Suggests` should work. 
https://github.com/keller-mark/anndataR/blob/d192e68ed312f212476a5490c973619f08e7c8de/R/ZarrAnnData.R#L240-L246

Artur-man committed:
add pizzarr to Suggests and README

lazappi commented:
Not a proper review of the code but I wanted to raise that we plan to submit **{anndataR}** to Bioconductor (or maybe CRAN) at some point (hopefully soonish). This means we won't be able to depend on GitHub only packages. Do you have plans to submit **{pizzarr}** somewhere or could you rewrite this to use one of the public Zarr packages?

Artur-man commented:
@lazappi thanks for the notification, `pizzarr` could be submitted to CRAN soon. https://github.com/keller-mark/pizzarr/pull/93

Artur-man commented:
@keller-mark any opinions on this ? perhaps we can make Rarr/pizzarr optional with `options()` depending on the speed. 

Regardless it would be good to have zarr support for AnnDataR. We are currently using `reticulate`/`basilisk` env in `SpatialData` and causes problems already (too many dependancies, multiple env etc.). https://github.com/HelenaLC/SpatialData

That being said, any dependancy should either be in CRAN or BioC. If there are missing utilities in `Rarr` we can also contact the maintainer and/or send PRs there (I did some, which I can continue too).

Artur-man committed:
proj

Artur-man committed:
Merge branch 'main' into keller-mark/zarr

Artur-man committed:
add keller-mark/pizzarr to Remotes

lazappi commented on inst/extdata/example.zarr/.zattrs:
Can we please zip or something the example to make it a single file? I don't think we want to be tracking ~400 extra files in `git`. If we can avoid needing an example file that might be even better but I'm happy to keep it if it's bundled in some way.

Artur-man replied:
should be easy to zip and unzip at tmpdir, will do

Artur-man replied:
done

Artur-man committed:
zip example.zarr

Artur-man committed:
Merge branch 'main' into keller-mark/zarr

Artur-man committed:
Merge branch 'main' into keller-mark/zarr

Artur-man committed:
adapt read_zarr to Rarr

Artur-man committed:
adapt write_zarr to Rarr

Artur-man commented:
I have quickly checked if `pizzarr` utilities could be replaced with https://github.com/grimbough/Rarr, unfortunately there exists a set of limitations to the BioC native package, which 
- [x] opening zarr stores and creating groups, see https://github.com/Huber-group-EMBL/Rarr/pull/18
- [x] read/write boolean arrays, see https://github.com/Huber-group-EMBL/Rarr/pull/59
- [x] fixing errors when reading from some character arrays, see https://github.com/grimbough/Rarr/issues/20
   
Will be in touch to see if these are resolved in the future, otherwise no zarr R package is currently both in CRAN/BioC and functionally complete yet.

Artur-man commented:
There is some progress in the Rarr package, would you guys like a clean PR (since there were so many updates since) or continuing here is fine ?

Artur-man committed:
update to most recent anndataR

lazappi commented:
> There is some progress in the Rarr package, would you guys like a clean PR (since there were so many updates since) or continuing here is fine ?

Probably whatever is easiest for you and @keller-mark/whatever makes the PR easiest to understand. There have been a lot of changes to the package since this was opened so we would need to make sure those are included here.

I saw that **{Rarr}** is planning to have Zarr v3 support for the next release so I think that makes sense in terms of which backend package to use.

Artur-man commented:
I adapted the changes, was fairly easy. I will then continue here and refer back to Hugo if needed again.

lazappi commented:
I'm going to make this a draft for now, just to help with our organisation. Please let us know when it's ready to review.

Artur-man committed:
remove old scripts

Artur-man committed:
update write_zarr

Artur-man committed:
initial update to ZarrAnnData

Artur-man committed:
update ZarrAnnData, documentation, and implement read_zarr_rec_array

Artur-man committed:
review read zarr helpers, and update tests

Artur-man committed:
update read_zarr, read tests pass

Artur-man committed:
some updates for writing zarr

Artur-man committed:
update write_empty_zarr

Artur-man committed:
remove pizzarr, update documentation

Artur-man committed:
remove pizzarr from tests

Artur-man committed:
fix test-ZarrAnnData

Artur-man committed:
update ZarrAnnData to imitate HDF5AnnData

Artur-man committed:
check redundant files, correct lines

Artur-man commented:
Guys, I am close .... I will add a new `example.zarr` here, however lets add a yml file to allow creating quick environments, so the same files can be generated quickly later, and also ... would you like a separate `example_zarr.py` file ? 

https://github.com/scverse/anndataR/blob/5ffd79a05bbc4a99194b89a6432777cf2f1760a5/inst/scripts/example_h5ad.py#L1-L7

Artur-man committed:
update example_h5ad.py, add zarr and change to example_files.py

lazappi commented:
We tried to cover as many cases as possible in the example dataset so I would probably modify the script to also output a (tarred/zipped) Zarr as well as an H5AD (and maybe rename it to something like `create_example.py`) rather than creating a new one.

I noted down the versions just in case it became important later but maybe it makes sense to store them in a separate file. If you are regenerating the dataset I would probably also update the environment. I think there is also a version number in the script that should be bumped and a changelog that should be updated.

Bisaloo commented on R/Rarr_utils.R:
```diff
@@ -0,0 +1,164 @@
+#' @param version zarr version
+#' @export
+create_zarr_group <- function(store, name, version = "v2"){
+  split.name <- strsplit(name, split = "\\/")[[1]]
```
It'll be slightly faster if regexes are not used:

```suggestion
  split.name <- strsplit(name, split = "/", fixed = TRUE)[[1]]
```

Artur-man replied:
done

Bisaloo commented on R/Rarr_utils.R:
```diff
@@ -0,0 +1,164 @@
+create_zarr_group <- function(store, name, version = "v2"){
+  split.name <- strsplit(name, split = "\\/")[[1]]
+  if(length(split.name) > 1){
+    split.name <- vapply(seq_len(length(split.name)),
```
```suggestion
    split.name <- vapply(seq_along(split.name),
```

Artur-man replied:
done

Artur-man committed:
add new test example

Artur-man committed:
some linting changes

Artur-man committed:
remove read/write_zattrs since implemented in Rarr

Artur-man committed:
access read/write_zarr_attr

Artur-man committed:
Merge branch 'main' into keller-mark/zarr

Artur-man committed:
add some missing tests

lazappi commented:
@Artur-man We have updated the CI. You will need to merge the `devel` branch here to get the new version.

Artur-man committed:
Merge branch 'main' into keller-mark/zarr

Artur-man committed:
update readers, update tests

Bisaloo commented on R/write_zarr_helpers.R:
```diff
@@ -0,0 +1,553 @@
+    dims <- dim(value)
+  } else {
+    dims <- length(value)
+  }
```
If you're using a recent version of R:

```suggestion
  dims <- dim(value) %||% length(value)
```

(you can also redefine the operator in this pkg for compatibility with older R versions)

Artur-man replied:
oh I will change this directly, anndataR requires 4.5 and later anyways. Thanks Hugo.

Artur-man committed:
correct nullable string zarr array write/read, introduce ordering in categorical zarr array

Artur-man committed:
do some linting, fix commented out code

Artur-man committed:
update some zarr writers and classes

Artur-man committed:
fix documentation

Artur-man committed:
fix compression interface for zarr

Artur-man committed:
full lint check

Artur-man committed:
fix examples

Artur-man committed:
check, biocheck and lintr

Artur-man committed:
fix development status

Artur-man committed:
air format

Artur-man committed:
air format test

Artur-man committed:
update example.zarr.zip, skip some test (waiting for Rarr)

Artur-man commented:
Some functions such as `read_h5ad_data_frame` and `read_h5ad_collection` would repeat themselves if we were to define a, e.g. `read_zarr_data_frame/_collection` since they do not directly call `rhdf5` or `Rarr` methods. 

I would instead open a new `read/write_file_helpers.R` to define these cross file format generic methods, and only keep file specific methods like `read_hdf5_numeric_scalar` where they are. 

If you want, I can implement them, otherwise we can explore after you review my changes. Up to you. I am still waiting for some changes in `Huber-group-EMBL/Rarr` but mostly we are ready.

Artur-man committed:
update example.zarr, fix some read_zarr_

Artur-man committed:
fix examples

Artur-man committed:
remove overwrite

Bisaloo reviewed (commented):
FYI, the patterns I highlight here can also be detected via lintr if you use `all_linters()` and then selectively turn off the linters you don't want.

For example: https://github.com/Huber-group-EMBL/Rarr/blob/devel/.lintr

Artur-man committed:
R code styling

lazappi commented:
> Some functions such as `read_h5ad_data_frame` and `read_h5ad_collection` would repeat themselves if we were to define a, e.g. `read_zarr_data_frame/_collection` since they do not directly call `rhdf5` or `Rarr` methods.

I think they are slightly different because one will call `read_h5ad_element()` and the other will call `read_zarr_element()`. I'm open to reducing redundancy where it makes sense and is still easy to follow but maybe it's better to leave for a separate PR so it's easier to review the Zarr functionality without other changes.

Artur-man commented:
Then @lazappi perhaps we can start your review, let me know if you guys want me to change anything (there are lots of changes!). Meanwhile, @Bisaloo will keep working on fixing a few small issues in `Rarr`.

lazappi commented on inst/scripts/example_files.py:
```diff
@@ -91,3 +102,12 @@
+# zip = zipfile.ZipFile("example.zarr.zip", "w", zipfile.ZIP_DEFLATED)
+# zip.write("example.zarr")
+# shutil.rmtree("example.zarr")
+# zip.close()
```
Maybe this should be uncommented?

Artur-man replied:
done

lazappi commented on inst/scripts/requirements.yml:
If we are adding this we can probably remove the versions from the script.

Bisaloo replied:
I learned about this recently, which can be a good lightweight way to document versions without having to create a dedicated environment file: https://docs.astral.sh/uv/guides/scripts/#declaring-script-dependencies

Note that it's not uv-specific but it's also [documented in the official python docs](https://packaging.python.org/en/latest/specifications/inline-script-metadata/#inline-script-metadata).

Artur-man replied:
Hmm, I actually like this: 
https://docs.astral.sh/uv/guides/scripts/#declaring-script-dependencies

Let me know what you think @lazappi, I can embed it and drop yml ... but I do not know if this would make development harder for people used to running conda env through yml.

lazappi replied:
Yeah, I think I prefer this to having a separate file as well

lazappi commented on R/AbstractAnnData.R:
```diff
@@ -287,6 +287,28 @@ AbstractAnnData <- R6::R6Class(
       )
     },
     #' @description
+    #' Convert to an [`ZarrAnnData`]
```
```suggestion
    #' Convert to a [`ZarrAnnData`]
```

Artur-man replied:
fixed

lazappi commented on R/AbstractAnnData.R:
```diff
@@ -287,6 +287,28 @@ AbstractAnnData <- R6::R6Class(
+    #' @param compression See [as_ZarrAnnData()]
+    #' @param mode See [as_ZarrAnnData()]
+    #'
+    #' @return An [`ZarrAnnData`] object
```
```suggestion
    #' @return A [`ZarrAnnData`] object
```

Artur-man replied:
fixed

lazappi commented on R/AbstractAnnData.R:
```diff
@@ -302,6 +324,23 @@ AbstractAnnData <- R6::R6Class(
       write_h5ad(object = self, path, compression = compression, mode = mode)
+    },
+    #' @description
+    #' Write the `AnnData` object to an H5AD file
```
```suggestion
    #' Write the `AnnData` object to a Zarr file
```

lazappi commented on R/Rarr_utils.R:
```diff
@@ -0,0 +1,75 @@
+#' store <- tempfile(fileext = ".zarr")
+#' create_zarr(store)
+#' create_zarr_group(store, "gp")
+#' @export
```
These are internal functions that shouldn't be exported, use `@noRd` instead

lazappi replied:
Please be more consistent with the formatting of the function documentation, something like:

```suggestion
#' Create Zarr group
#'
#' Create a new group in a Zarr store
#'
#' @param store the location of the Zarr store
#' @param name name of the group to create
#' @param version Zarr version
#'
#' @noRd
#'
#' @examples
#' store <- tempfile(fileext = ".zarr")
#' create_zarr(store)
#' create_zarr_group(store, "gp")
```

lazappi commented on R/Rarr_utils.R:
```diff
@@ -0,0 +1,75 @@
+#' @param store the location of (zarr) store
+#' @param name name of the group
+#' @param version zarr version
+#' @importFrom utils tail
```
Imports from common packages can go in `anndataR-package.R`

Artur-man replied:
fixed

lazappi commented on R/Rarr_utils.R:
```diff
@@ -0,0 +1,75 @@
+#' @export
+#' @return `NULL`
+create_zarr_group <- function(store, name, version = "v2") {
+  split.name <- strsplit(name, split = "/", fixed = TRUE)[[1]]
```
Use `snake_case` for variable names

Artur-man replied:
ok, I think it is fixed.

lazappi commented on R/Rarr_utils.R:
```diff
@@ -0,0 +1,75 @@
+      )
+    },
+    v3 = {
+      stop("Currently only zarr v2 is supported!")
```
Use `cli_abort()` for errors

Artur-man replied:
fixed

lazappi commented on R/Rarr_utils.R:
```diff
@@ -0,0 +1,75 @@
+        file = file.path(store, split.name[1], ".zgroup")
+      )
+    },
+    v3 = {
```
What are your thoughts on being able to support v3?

Artur-man replied:
@Bisaloo is working on it, i think he wants to do this until march-april

Artur-man replied:
We can read zarr v3 now ... but not writing, perhaps this is enough for pushing zarr support to anndataR release.

lazappi replied:
Ok. I think the way this is set up is reasonable. The argument makes it clear we know there are versions and the message should be obvious that only v2 is currently supported. I would adjust the messages a bit but I will suggest something later.

lazappi replied:
I think we should delay adding v3 write support to a future PR (when all the required functionality is available)

lazappi commented on R/Rarr_utils.R:
```diff
@@ -0,0 +1,75 @@
+    }
+  }
+  dir.create(file.path(store, split.name[1]), showWarnings = FALSE)
+  switch(
```
Should this check be in `create_zarr()` (or wherever it will be first triggered)?

Artur-man replied:
Hmm, good idea but I think it doesn't matter where it is put. Either in `create_zarr` or `create_zarr_group` this check should be performed. Since root zarr is a zarr group too, we call `create_zarr_group` with `create_zarr`. 

However, I like how you think. This functions are auxiliary and will be deprecated once @Bisaloo implements it in `Rarr`.

Artur-man replied:
I would keep the check for zarr format here (until we can write zarr groups/arrays directly). Since zarr store is already a zarr root group and this check should be done for groups too. 

That being said, since we do not define higher-level classes for zarr stores (and instead open/create/modify groups and arrays directly via path), we cannot check if, for example, if the original zarr store is v3 or while the group you are trying to create is v2, which should normally throw an error. 

However, all these requires major changes to Rarr and its inner workings.

lazappi commented on R/Rarr_utils.R:
```diff
@@ -0,0 +1,75 @@
+#' create_zarr(store)
+#' is_zarr_empty(store)
+#' @export
+#' @return returns TRUE if zarr store is not empty
```
If the function is named `is_zarr_empty()` it should return `TRUE` when the store is empty. I'm not sure if it's the name or the documentation that is wrong but they don't match.

Artur-man replied:
fixed

lazappi commented on R/read_zarr_helpers.R:
```diff
@@ -0,0 +1,521 @@
+      if (stop_on_error) {
+        stop("No encoding info found for element '", name, "'")
+      } else {
+        warning("No encoding found for element '", name, "'")
```
Use `cli_warn()` for warnings

Artur-man replied:
fixed

lazappi commented on R/read_zarr_helpers.R:
```diff
@@ -0,0 +1,521 @@
+#' @return A named list with names type and version
+#'
+#' @noRd
+read_zarr_encoding <- function(store, name, stop_on_error = TRUE) {
```
The `stop_on_error` isn't needed here, please make it match the `read_h5ad_encoding()` function

Artur-man replied:
fixed

lazappi commented on R/read_zarr_helpers.R:
```diff
@@ -0,0 +1,521 @@
+      name,
+      stop_on_error = stop_on_error
+    )
+    if (is.null(encoding_list)) {
```
This check isn't needed because `read_zarr_encoding()` should fail if there is no encoding found

Artur-man replied:
fixed

lazappi commented on R/read_zarr_helpers.R:
```diff
@@ -0,0 +1,521 @@
+    dims <- dim(data)
+    data <- as.logical(data)
+    dim(data) <- dims
+  }
```
This probably isn't needed here

lazappi commented on R/read_zarr_helpers.R:
```diff
@@ -0,0 +1,521 @@
+  if (is.null(dim(data))) {
+    data <- as.vector(data)
+    dim(data) <- length(data)
+  }
```
This may not be needed depending on what `Rarr::read_zarr_array()` returns

Bisaloo replied:
`Rarr::read_zarr_array()` currently always returns arrays (i.e., `dim()` is not `NULL`). I don't expect this to change as I like the idea of a consistent output class.

lazappi commented on R/read_zarr_helpers.R:
```diff
@@ -0,0 +1,521 @@
+      as.vector(Rarr::read_zarr_array(file.path(store, name, x)))
+    }),
+    field_names
+  )
```
```suggestion
  purrr::map(field_names, function(.field) {
    as.vector(read_zarr_array(store, file.path(name, .field)))
  }) |>
    setNames(field_names)
```

Artur-man replied:
lets leave this for now, because we are still waiting @Bisaloo to implement readers for structured data. 
https://github.com/Huber-group-EMBL/Rarr/issues/81

lazappi commented on R/read_zarr_helpers.R:
```diff
@@ -0,0 +1,521 @@
+      as.vector(Rarr::read_zarr_array(file.path(store, name, x)))
+    }),
+    field_names
+  )
```
Maybe this could use `read_zarr_collection()`?

Artur-man replied:
will leave until https://github.com/Huber-group-EMBL/Rarr/issues/81 is fixed.

lazappi commented on R/read_zarr_helpers.R:
```diff
@@ -0,0 +1,521 @@
+  }
+
+  # convert "NA" to NA (as in rhdf5:::.h5postProcessDataset)
+  data[data == "NA"] <- NA
```
I don't think `string_array` is nullable but we should check

Artur-man replied:
This we have discussed with @Bisaloo to how to implement. Currently, rhdf5 forces NAs to "NA" when saved as string arrays, and then return as NA back when returned from the `.h5`.

This doesnt happen in `Rarr`, so we imitate in `anndataR` instead. 

See more here: https://github.com/Huber-group-EMBL/Rarr/issues/86

lazappi commented on R/read_zarr_helpers.R:
```diff
@@ -0,0 +1,521 @@
+    stop(
+      "There is currently no support for multidimensional categorical arrays"
+    )
+  }
```
I don't think this is needed as this can never be written

Artur-man replied:
fixed

lazappi commented on R/read_zarr_helpers.R:
```diff
@@ -0,0 +1,521 @@
+    )
+
+    ordered <- FALSE
+  }
```
I don't think this is needed as the attribute should always be written

Artur-man replied:
fixed

lazappi commented on R/read_zarr_helpers.R:
```diff
@@ -0,0 +1,521 @@
+#' If `include_index == TRUE` the index stored in the Zarr store is added as a
+#' column to output `data.frame` using the defined index name as the column
+#' name and this is set as an attribute. If `include_index == FALSE` the index
+#' is not provided in the output. In either case row names are not set.
```
This should have the same behaviour as `read_h5ad_data_frame()`. The index should always be read and set as the row names. I think the function is correct but the documentation needs updating.

Artur-man replied:
fixed

lazappi commented on R/write_zarr.R:
```diff
@@ -0,0 +1,122 @@
+#' Write Zarr
+#'
+#' Write an Zarr file
```
```suggestion
#' Write a Zarr file
```

Artur-man replied:
fixed

lazappi commented on R/write_zarr.R:
```diff
@@ -0,0 +1,122 @@
+#' ## Compression
+#'
+#' Compression is currently not supported for Boolean arrays, they will be
+#' written uncompressed.
```
Is this true for Zarr as well?

lazappi commented on R/read_zarr_helpers.R:
```diff
@@ -0,0 +1,521 @@
+  }
+
+  read_fun <- switch(
+    type,
```
Need to add a `read_zarr_null()` function

Artur-man replied:
fixed, @lazappi are there tests for this for hdf5/h5ad ?

lazappi commented on R/write_zarr_helpers.R:
```diff
@@ -0,0 +1,693 @@
+  compression <- match.arg(compression)
+
+  # Sparse matrices
+  write_fun <-
```
Need to add `write_zarr_null()`

Artur-man replied:
need to check with @Bisaloo for this.

Artur-man replied:
the read and write of null encoding is strictly based on the `.zattrs`. I have tried reading the null array in a zarr v2 implementation and that gives an error. Thus, I will just write an empty logical array here with null encoding in the attr. I will leave a `TODO:` for future correction though.

lazappi replied:
We should check this matches whatever Python does

lazappi commented on R/write_zarr_helpers.R:
```diff
@@ -0,0 +1,693 @@
+) {
+  dims <- dim(value) %||% length(value)
+
+  # replace NA to "NA" (as in rhdf5:::.h5postProcessDataset)
+  # to read as "NA" -> NA later after Rarr:read_zarr_array
+  value[is.na(value)] <- "NA"
```
Don't think we need this

Artur-man replied:
nope, unfortunately the behaviour is not the same as in rhdf5. @Bisaloo wanna comment ? 

```
> store <- tempfile(fileext = ".zarr")
> create_zarr(store = store)
> string <- LETTERS[1:5]
> string[5] <- NA
> write_zarr_element(string, store, "string_array")
NULL
Warning message:
Could not write element 'string_array' of type 'character': missing value where TRUE/FALSE needed 
```

lazappi replied:
The main thing is not that our Zarr and H5AD backends are the same but that they are both consistent with their matching Python implementations

lazappi replied:
I think this behaviour might actually be incorrect for H5AD, I've created an issue to follow up https://github.com/scverse/anndataR/issues/444

lazappi commented on R/ZarrAnnData.R:
```diff
@@ -0,0 +1,497 @@
+
+      is_readonly <- FALSE
+
+      if (is.character(file)) {
```
I think for Zarr you only have a path, no handle object? If that is the case some of the logic here can be simplified.

Artur-man replied:
ok, I added some additional checks, thus now if `file` is not a character then it will give an error. But let me know I have missed something here. 

I also check if the file is a valid zarr folder, which is different than the folder being a valid anndata-zarr.

lazappi commented on R/ZarrAnnData.R:
```diff
@@ -0,0 +1,497 @@
+      attrs <- Rarr::read_zarr_attributes(file)
+      if (!all(c("encoding-type", "encoding-version") %in% names(attrs))) {
+        cli_abort(c(
+          "File {.file {file}} is not a valid Zarr file."
```
Maybe this (and similar messages/documentation) needs to say something like "AnnData Zarr file" because there is no AnnData-specific extension like there is for H5AD.

Artur-man replied:
I see makes sense. Fixed

lazappi commented on R/ZarrAnnData.R:
```diff
@@ -0,0 +1,497 @@
+
+    # We don't close
+    #' @description Close the Zarr store/file
+    close = function() {},
```
This can be removed if it's not needed

Artur-man replied:
fixed

lazappi commented on R/ZarrAnnData.R:
```diff
@@ -0,0 +1,497 @@
+}
+
+# nolint start: object_name_linter
+cleanup_ZarrAnnData <- function(...) {
```
I don't think this is used anywhere? It can be removed if so.

Artur-man replied:
fixed

lazappi commented on tests/testthat/test-h5ad-zarr.R:
```diff
@@ -0,0 +1,125 @@
+})
+
+test_that("reading recarrays works", {
+  skip("read_zarr_rec_array is not implemented yet")
```
I think this is implemented now?

Artur-man replied:
Hmm, ok the skip here is misleading, since `Rarr` currently doesn't read arbitrary structured data types, but my current implementation under `read_zarr_rec_array` only works for the example data. So it should stay like this until the following is fixed: 
https://github.com/Huber-group-EMBL/Rarr/issues/81

Artur-man replied:
done

lazappi commented on tests/testthat/test-h5ad-zarr.R:
```diff
@@ -0,0 +1,125 @@
+  expect_equal(mapping_h5ad, mapping_zarr)
+})
+
+test_that("reading dataframes works", {
```
```suggestion
test_that("reading dataframes is the same for h5ad and zarr", {
```

Artur-man replied:
fixed

lazappi commented on tests/testthat/test-h5ad-zarr.R:
```diff
@@ -0,0 +1,125 @@
+})
+
+test_that("reading H5AD as SingleCellExperiment is same for h5ad and zarr", {
+  skip("for now, example.zarr and example.h5ad are not identical!")
```
Why are they not identical?

Artur-man replied:
some readers are not implemented yet, but will check again.
https://github.com/scverse/anndataR/pull/190#issuecomment-3619051125

Artur-man replied:
should be done now

lazappi commented on tests/testthat/test-h5ad-zarr.R:
```diff
@@ -0,0 +1,125 @@
+  expect_equal(df_h5ad, df_zarr)
+})
+
+test_that("reading H5AD as SingleCellExperiment is same for h5ad and zarr", {
```
Should probably add a test for Seurat

Artur-man replied:
fixed, but skipping for now until identical.

Artur-man replied:
just letting you know, exact equality here is hard to achieve ... this is because of how structed data types (rec arrays are read between h5ad and zarr, and even between h5py and rhdf5):

see here
Huber-group-EMBL/rhdf5#192

Artur-man replied:
mostly done, but also being addressed in https://github.com/scverse/anndataR/issues/409

lazappi commented on vignettes/articles/software_design.Rmd:
```diff
@@ -50,7 +50,7 @@ The different `AnnData` classes provide a consistent interface but store and acc
 - The `ReticulateAnnData` access data stored in an `AnnData` object in a concurrent Python session.
   This comes with the overhead and complexity of using `r CRANpkg("reticulate")` but is sometimes useful to access functionality that has not yet been implemented in `r Biocpkg("anndataR")`.
-- The planned `ZarrAnnData` will provide an interface to an `AnnData` Zarr store, similar to `HDF5AnnData`.
+- The `ZarrAnnData` provides an interface to a Zarr data (i.e. Zarr store) and, similar to a H5AD file, minimal data is stored in memory until it is requested by the user. It is as an intermediate object when reading/writing Zarr store but can be useful for accessing parts of large files.
```
```suggestion
- The `ZarrAnnData` provides an interface to a Zarr store and, similar to a H5AD file, minimal data is stored in memory until it is requested by the user. It is as an intermediate object when reading/writing Zarr stores but can be useful for accessing parts of large files.
```

Artur-man replied:
fixed

lazappi reviewed (changes requested):
We will probably need some more rounds of reviews but hopefully I've caught a lot of stuff already.

Additional comments:
- I commented on some style things the first time I saw them but please do the same wherever else it is relevant
- The documentation in `AnnData-usage` needs to be updated to add `ZarrAnnData`
- Round trip tests for Zarr need to be added
- We should probably show Zarr usage somewhere in a main vignette
- Please try to get the CI checks to pass

Artur-man commented:
> We will probably need some more rounds of reviews but hopefully I've caught a lot of stuff already.
> 
> Additional comments:
> 
> * I commented on some style things the first time I saw them but please do the same wherever else it is relevant
> * The documentation in `AnnData-usage` needs to be updated to add `ZarrAnnData`
> * Round trip tests for Zarr need to be added
> * We should probably show Zarr usage somewhere in a main vignette
> * Please try to get the CI checks to pass

awesome @lazappi I will get to it.

Bisaloo commented on README.md:
```diff
@@ -54,6 +54,9 @@ the task you want to perform.
 - To convert to/from `Seurat` objects, install [SeuratObject](https://cran.r-project.org/package=SeuratObject):  
   `install.packages("SeuratObject")`
+- To read/write \*.zarr files, you need to install
+  [zarr](https://www.bioconductor.org/packages/release/bioc/html/Rarr.html):  
```
```suggestion
  [Rarr](https://www.bioconductor.org/packages/release/bioc/html/Rarr.html):  
```

Artur-man replied:
fixed

Artur-man committed:
fixes from @lazappi

Artur-man committed:
Merge branch 'main' into keller-mark/zarr

Artur-man committed:
air format

Artur-man committed:
update some documentation

Artur-man committed:
fix some tests

Artur-man commented:
Errors I am getting at lint and pkgdown GHAs look unrelated to the package: 

```
Error in dyn.load(file, DLLpath = DLLpath, ...) : 
    unable to load shared object '/home/runner/work/_temp/Library/Rhdf5lib/libs/Rhdf5lib.so':
    libsz.so.2: cannot open shared object file: No such file or directory
```

Artur-man commented:
@Bisaloo also the tests that are passing for macos, ubuntu and windows are failing for macos (intel) and ubuntu-bioc-release. I understand the error in release (for obvious reasons), but macos-15 I don't know. Any ideas ? 

```
Running the tests in ‘tests/testthat.R’ failed.
Last 13 lines of output:
   5.         └─base::array(metadata$fill_value, dim = lengths(index))
  ── Error ('test-Zarr-read.R:87:3'): reading numeric scalars works ──────────────
  Error in `array(metadata$fill_value, dim = lengths(index))`: 'dim' cannot be of length 0
  Backtrace:
      ▆
   1. └─anndataR:::read_zarr_numeric_scalar(store, "uns/IntScalar") at test-Zarr-read.R:87:3
   2.   └─anndataR:::read_zarr_array(store, name)
   3.     └─Rarr::read_zarr_array(file.path(store, name))
   4.       └─Rarr:::read_data(...)
   5.         └─base::array(metadata$fill_value, dim = lengths(index))
  
  [ FAIL 4 | WARN 447 | SKIP 12 | PASS 1842 ]
```

https://github.com/scverse/anndataR/actions/runs/20411185095/job/58648259569?pr=190

Bisaloo commented:
Yes, it looks that on this runner, Rarr 1.11.9 is installed, when support for reading scalars was added in Rarr 1.11.12. I think it's related to BioC devel not having more recent binaries for macOS x86 yet.

Maybe try setting:

```yaml
Imports:
    Rarr (>= 1.11.12)
```

In `DESCRIPTION` to force installation from source?

This will not work for the runner using BioC release but it's already failing anyways.

Artur-man commented:
I see. Then I will assume all checks are passing for now. thanks man

Artur-man commented:
Current issues: 

- [x] https://github.com/Huber-group-EMBL/Rarr/pull/82
- [x] https://github.com/Huber-group-EMBL/Rarr/issues/86
- [x] https://github.com/Huber-group-EMBL/Rarr/issues/81
- [x] implement `write_zarr_null`.
- [x] update `ZarrAnnData$initialize` since Zarr store doesn't have a handler.
- [x] The documentation in AnnData-usage needs to be updated to add ZarrAnnData
- [x] We should probably show Zarr usage somewhere in a main vignette
- [x] https://github.com/Huber-group-EMBL/Rarr/pull/104
- [x] Support for Unicode rec-arrays https://github.com/Huber-group-EMBL/Rarr/issues/81 
- [x] fix roundtrip tests for zarr
- [x] remove `eval=FALSE` from vignette chunks after fixes
- [x] Get the CI checks to pass

Artur-man committed:
more fixes on anndata-zarr integration

Artur-man committed:
update ZarrAnnData$initialize

Artur-man committed:
update zarr compression

Artur-man committed:
fix column-order here, C based ordering for arrays

Artur-man committed:
implement roundtrip tests for anndata-zarr

Artur-man committed:
add zarr to vignettes

Artur-man committed:
update README and software_design.rmd

Artur-man committed:
update AnnData-usage

Artur-man committed:
update write_zarr documentation

Artur-man committed:
update write_zarr_null

Artur-man committed:
fix rec_array, update tests and example datasets

Artur-man committed:
fix duplicate chunks in Rmd

Artur-man committed:
add write_zarr_null

Artur-man committed:
update write string array (zarr), air and lint

Artur-man committed:
implement writing empty zarr elements

Artur-man committed:
update tests for rec_array conformance of h5ad and zarr

Artur-man committed:
update mapping conformance test for h5ad and zarr

Artur-man committed:
implement H5_ITER like ordering and fix h5ad vs zarr testing

Artur-man committed:
air and lint

Artur-man committed:
fix test bug

Artur-man committed:
do not call expect_equal outside of test

Artur-man commented on tests/testthat/test-h5ad-zarr.R:
```diff
@@ -0,0 +1,169 @@
+  expect_equal(array_h5ad, array_zarr)
+})
+
+test_that("reading mappings is same for h5ad and zarr", {
```
There is a difference between rhdf5, Rarr and h5py on how rec arrays and structured data types are read. 

To me, h5py and Rarr reads it correctly, but rhdf5 does not:
Huber-group-EMBL/rhdf5#192

lazappi replied:
Maybe we need an issue for this here as well? From our point of view it doesn't matter so much what is returned by **{rhdf5}** as long as what we return from the `AnnData` matches Python.

Going the other way (writing a `recarray`) is more difficult because we don't know when to use that format. I don't think we test this. Also, I don't think the `AnnData` disk spec really covers alternative arrays like `recarray`.

Artur-man replied:
https://github.com/scverse/anndataR/issues/409

Artur-man commented:
We are really close, tests are passing on my local but we have to wait until changes in `Rarr` are in BioC-devel. 
Let me give a break I will review the entire PR. Please also let me know if you catch stuff, otherwise I will get back to this to finalize. 

Thanks @Bisaloo again for answering all issues of Rarr!

Bisaloo commented:
As discussed in the bioconductor zulip, you're not likely to get up-to-date macOS x86 binaries for Rarr. So you'll need to either force installing from source, or add the bioc r-universe as an additional repository for this test to pass.

Alternatively, setting 1.11.24 as minimum required version for Rarr might do the trick to force installing from source when a recent enough binary isn't available.

Artur-man committed:
implement examples, test and datasets for zarr v3

Artur-man committed:
Merge branch 'main' into keller-mark/zarr

Artur-man committed:
fix issues, lint and update example datasets to new anndata version

Artur-man committed:
Merge branch 'main' into keller-mark/zarr

lazappi commented:
> As discussed in the bioconductor zulip, you're not likely to get up-to-date macOS x86 binaries for Rarr. So you'll need to either force installing from source, or add the bioc r-universe as an additional repository for this test to pass.
> 
> Alternatively, setting 1.11.24 as minimum required version for Rarr might do the trick to force installing from source when a recent enough binary isn't available.

Could you please explain what the issue with this is? Is it something on the Bioconductor side so do we need to look into it?

I'm guessing the current release check is also expected to fail because we need a later version of **{Rarr}**?

Bisaloo commented:
> Could you please explain what the issue with this is? Is it something on the Bioconductor side so do we need to look into it?

BioC macOS x86 builder is no longer working. They're planning on pulling the binaries from r-universe, as they are doing with windows binaries, but it's still a work in progress.
In the meantime, the latest available version for macOS x86 binaries in BioC devel is quite outdated.
It's something that needs to be (and will be) resolved by Bioconductor.

But you can still work around the issue if you need to. By requiring installation from source (this should happen automatically if you request a version more recent than what the binaries provide). Or by getting the binaries from r-universe yourself.

> I'm guessing the current release check is also expected to fail because we need a later version of {Rarr}?

Yes, exactly

rcannood commented on inst/extdata/example_v2.zarr.zip:
Can we Rbuildignore the example_v2.zarr.zip and example_v3.zarr.zip files?

Artur-man replied:
what about example.h5ad ?

lazappi replied:
We use these in the examples/tests so they need to be in the package

rcannood commented on inst/scripts/requirements.yml:
```diff
@@ -0,0 +1,14 @@
+      - scipy==1.17.1
+      - zarr==3.1.5
+      - igraph==1.0.0
+      - leidenalg==0.11.0
```
```suggestion
      - leidenalg==0.11.0

```

Add missing newline

Artur-man replied:
done

rcannood commented on R/AbstractAnnData.R:
```diff
@@ -313,6 +335,37 @@ AbstractAnnData <- R6::R6Class(
+      mode = c("w-", "r", "r+", "a", "w", "x")
+    ) {
+      write_zarr(
+        object = self, 
```
fix linting issues

rcannood commented on R/AbstractAnnData.R:
```diff
@@ -313,6 +335,37 @@ AbstractAnnData <- R6::R6Class(
+    #' @return `path` invisibly
+    write_zarr = function(
+      path,
+      compression = c(
```
the compression arguments of `write_zarr` is different from `as_ZarrAnnData`. which one is correct?

Artur-man replied:
are you talking about the possible values of `compression`. I think they are same already? 

```
as_ZarrAnnData <- function(
  # nolint end: object_name_linter
  adata,
  file,
  compression = c(
    "none",
    "gzip",
    "blosc",
    "zstd",
    "lzma",
    "bz2",
    "zlib",
    "lz4"
  ),
  mode = c("w-", "r", "r+", "a", "w", "x")
```

```
write_zarr <- function(
  object,
  path,
  compression = c(
    "none",
    "gzip",
    "blosc",
    "zstd",
    "lzma",
    "bz2",
    "zlib",
    "lz4"
  ),
  mode = c("w-", "r", "r+", "a", "w", "x"),
  ...

lazappi replied:
This should be fixed now

rcannood commented on R/read_h5ad_helpers.R:
```diff
@@ -179,7 +179,6 @@ read_h5ad_element_keys <- function(
 #' @noRd
 read_h5ad_null <- function(file, name, version = "0.1.0") {
   version <- match.arg(version)
-
   NULL
```
undo this unnecessary change to avoid merge conflicts

Artur-man replied:
done

rcannood commented on tests/testthat/test-roundtrip-empty-zarr.R:
merge `test-roundtrip-*-zarr.R` into `test-roundtrip-*.R`

rcannood commented on vignettes/articles/software_design.Rmd:
```diff
@@ -23,34 +23,30 @@ knitr::opts_chunk$set(
 * `r Biocpkg("zellkonverter")` ([theislab/zellkonverter](https://github.com/theislab/zellkonverter)): Convert `AnnData` files to/from `SingleCellExperiment` objects
 * `r Githubpkg("mtmorgan/h5ad")`: Read/write `*.h5ad` files natively using `r Biocpkg("rhdf5")`
-* `r CRANpkg("anndataR")` ([dynverse/anndata](https://github.com/dynverse/anndata)): An R implementation of the `AnnData` data structures, uses `r CRANpkg("reticulate")` to read/write `*.h5ad` files
+* `r Biocpkg("Rarr")`: Read/write `*.zarr` stores natively
```
Rarr is a dependency, but is not a package related to reading or writing anndata files

Artur-man replied:
Hmm, ok you are right, so delete then ? I cannot think of an implementation to be inspired from based on `Rarr`.

rcannood reviewed (commented):
First round of comments prior to the hackathon!

Please merge devel again

Artur-man committed:
Merge branch 'main' into keller-mark/zarr

Artur-man committed:
lint and merge

Artur-man committed:
revert some lines

Artur-man committed:
small changes

Artur-man committed:
revert small changes

Artur-man committed:
air format some tests

lazappi committed:
Merge remote-tracking branch 'origin/devel' into keller-mark/zarr
* origin/devel:
  Bump actions/github-script from 8 to 9 (#443)
  Bump dawidd6/action-download-artifact from 19 to 20 (#440)
  Bump bencherdev/bencher from 0.6.0 to 0.6.2 (#442)
  Fix software design vignette (#438)
  Bump bencherdev/bencher from 0.5.10 to 0.6.0 (#439)
  Bump codecov/codecov-action from 5 to 6 (#437)
  Bump dawidd6/action-download-artifact from 16 to 19 (#436)

lazappi committed:
Set v2 in write_zarr_* helpers

lazappi committed:
Fix stop message in write_zarr_element()

lazappi committed:
Fix roxygen comment in write_zarr_element()

lazappi committed:
Expand compression list in as_ZarrAnnData()

lazappi committed:
Fix H5_ITER_INC_ORDERING docs

lazappi committed:
Fix as_ZarrAnnData() compression docs

lazappi committed:
Fix Zarr varm roundtrip test

lazappi committed:
Review duplicate entry in README

lazappi committed:
Fix typo in AnnData-usage docs

lazappi committed:
Fix comma in software design vignette

lazappi committed:
Adjust class descriptions in software design vignette

lazappi committed:
Roxygenise

lazappi commented on inst/scripts/example_files.py:
```diff
@@ -91,3 +108,25 @@
+
+# Write Zarr files in both v2 and v3 formats and zip them
+os.chdir("inst/extdata/")
+for fmt in (2, 3):
+    anndata.settings.zarr_write_format = fmt
+
+    zarr_dir = f"example_v{fmt}.zarr"
+    zip_path = f"{zarr_dir}.zip"
+
+    adata.write_zarr(zarr_dir)
+
+    with zipfile.ZipFile(zip_path, "w", zipfile.ZIP_DEFLATED) as z:
+        for root, dirs, files in os.walk(zarr_dir):
+            for file in files:
+                full_path = os.path.join(root, file)
+                arcname = os.path.relpath(
+                    full_path,
+                    start=os.path.dirname(zarr_dir)
+                )
+                z.write(full_path, arcname)
+
+    shutil.rmtree(zarr_dir)
```
Can this be simplified?

Bisaloo replied:
FYI, I'm planning to write a separate, more full-featured, package to handle Zarr stores after the next Bioconductor release and it'll handle zip stores out of the box, which could make you life a bit easier here.

lazappi commented on R/Rarr_utils.R:
```diff
@@ -0,0 +1,111 @@
+#'
+#' @param store the location of (zarr) store
+#' @param name name of the group
+#' @param version zarr version
```
Parameter descriptions should use sentence case e.g. `@param param Param does this`

lazappi commented on R/Rarr_utils.R:
```diff
@@ -0,0 +1,111 @@
+#' @examples
+#' store <- tempfile(fileext = ".zarr")
+#' create_zarr(store)
+#' create_zarr_group(store, "gp")
```
Don't need examples on unexported functions

lazappi commented on R/Rarr_utils.R:
```diff
@@ -0,0 +1,111 @@
+#' @noRd
+#'
+#' @examples
+#' store <- tempfile(fileext = ".zarr")
+#' create_zarr(store)
+#' create_zarr_group(store, "gp")
+create_zarr_group <- function(store, name, version = "v2") {
+  split_name <- strsplit(name, split = "/", fixed = TRUE)[[1]]
+  if (length(split_name) > 1) {
+    split_name <- vapply(
+      seq_along(split_name),
+      function(x) paste(split_name[seq_len(x)], collapse = "/"),
+      FUN.VALUE = character(1)
+    )
+    split_name <- rev(tail(split_name, 2))
+    if (!dir.exists(file.path(store, split_name[2]))) {
+      create_zarr_group(store = store, name = split_name[2])
+    }
+  }
+  dir.create(file.path(store, split_name[1]), showWarnings = FALSE)
```
Add comments explaining this

lazappi commented on R/read_zarr_helpers.R:
```diff
@@ -0,0 +1,481 @@
+  indptr <- as.vector(Rarr::read_zarr_array(file.path(
+    store,
+    paste0(name, "/indptr")
+  )))
```
Fix identation

lazappi commented on R/read_zarr_helpers.R:
```diff
@@ -0,0 +1,481 @@
+      repr = "R",
+      index1 = FALSE
+    )
+  }
```
Change to how this is done for H5AD. Could move to a function.

lazappi commented on R/read_zarr_helpers.R:
```diff
@@ -0,0 +1,481 @@
+  )
+
+  # Omit Zarr metadata files from the list of columns.
+  items <- items[!items %in% c(".zgroup", ".zattrs", ".zarray")]
```
Maybe we should store the metadata file vector somewhere and reuse it?

lazappi commented on R/read_zarr_helpers.R:
```diff
@@ -0,0 +1,481 @@
+  items <- items[!items %in% c(".zgroup", ".zattrs", ".zarray")]
+
+  # h5ad like ordering, see H5_ITER_INC
+  items <- H5_ITER_INC_ORDERING(items)
```
Why is this needed?

lazappi commented on R/write_zarr_helpers.R:
```diff
@@ -0,0 +1,728 @@
+    "zlib" = Rarr::use_zlib(),
+    "lz4" = Rarr::use_lz4()
+  )
+}
```
Document this, use a better variable name

lazappi commented on R/ZarrAnnData.R:
```diff
@@ -0,0 +1,501 @@
+#'
+#' See [AnnData-usage] for details on creating and using `AnnData` objects.
+#'
+#' @return An `ZarrAnnData` object
```
Use "A `ZarrAnnData`" not "An `ZarrAnnData`"

```suggestion
#' @return A `ZarrAnnData` object
```

lazappi commented on R/ZarrAnnData.R:
```diff
@@ -0,0 +1,501 @@
+      if (!zarr_path_exists(private$.zarrobj, "/")) {
+        cli_abort(
+          paste(
+            "The Zarr file does not exist, or not a zarr file/store!"
```
```suggestion
            "The Zarr path does not exist or is not a valid Zarr store"
```

lazappi commented on tests/testthat/test-h5ad-zarr.R:
Restructure this to use a similar approach as the roundtrip tests

lazappi commented on tests/testthat/test-Zarrv3-read.R:
Merge this into `test-Zarr-read.R`

lazappi commented on vignettes/articles/software_design.Rmd:
```diff
@@ -30,15 +30,11 @@ Ideally, this package will be a complete replacement for all of these packages,
 * An `r CRANpkg("R6")` `AnnData` class to work with objects in R
-* In-memory (`InMemoryAnnData`), HDF5-backed (`HDF5AnnData`) and Python-backed (`ReticulateAnnData`) back ends with a consistent interface
-* Read/write `.h5ad` files natively
+* In-memory (`InMemoryAnnData`), HDF5-backed (`HDF5AnnData`), Zarr-backed (`ZarrAnnData`), Python-backed (`ReticulateAnnData`) back ends with a consistent interface
```
```suggestion
* In-memory (`InMemoryAnnData`), HDF5-backed (`HDF5AnnData`), Zarr-backed (`ZarrAnnData`) and Python-backed (`ReticulateAnnData`) back ends with a consistent interface
```

lazappi commented on vignettes/anndataR.Rmd:
```diff
@@ -36,9 +36,9 @@ This package builds on our experience developing and using other interoperabilit
-  In contrast, `r Biocpkg("anndataR")` provides a native R H5AD interface, removing the need for Python dependencies.
+- `r Biocpkg("zellkonverter")` provides conversion of `SingleCellExperiment` objects to/from `AnnData` and reading/writing of `.h5ad` files and `.zarr` stores.
+  This is facilitated via `r CRANpkg("reticulate")` using `r Biocpkg("basilisk")` to manage Python environments (native reading of `.h5ad` and `.zarr` is also possible).
+  In contrast, `r Biocpkg("anndataR")` provides a native R H5AD and Zarr interface, removing the need for Python dependencies.
```
```suggestion
- `r Biocpkg("zellkonverter")` provides conversion of `SingleCellExperiment` objects to/from `AnnData` and reading/writing of `.h5ad`.
  This is facilitated via `r CRANpkg("reticulate")` using `r Biocpkg("basilisk")` to manage Python environments (native reading of `.h5ad` is also possible).
  In contrast, `r Biocpkg("anndataR")` provides a native R H5AD and Zarr interface, removing the need for Python dependencies.
```

lazappi commented on vignettes/anndataR.Rmd:
````diff
@@ -97,6 +97,31 @@ There is also a HDF5-backed `AnnData` object:
+Similarly, these functionalities are provided for `.zarr` stores too. 
+
+```{r zarr-path, eval = FALSE}
+library(anndataR)
+
+zarr_path <- system.file("extdata", "example.zarr.zip", package = "anndataR")
+td <- tempdir(check = TRUE)
+unzip(zarr_path, exdir = td)
+zarr_path <- file.path(td, "example.zarr")
+```
+
+```{r read-zarr, eval = FALSE}
+# in-memory
+adata <- read_zarr(zarr_path)
+
+# as SingleCellExperiment
+sce <- read_zarr(zarr_path, as = "SingleCellExperiment")
+
+# as SingleCellExperiment
+obj <- read_zarr(zarr_path, as = "Seurat")
+
+# as Zarr-backed
+adata <- read_zarr(zarr_path, as = "ZarrAnnData")
+```
+
````
These examples should be evaluated

lazappi commented on vignettes/anndataR.Rmd:
````diff
@@ -187,6 +212,19 @@ tmpfile <- tempfile(fileext = ".h5ad")
+tmpfile <- tempfile(fileext = ".zarr")
+write_zarr(obj, tmpfile)
+```
+
````
Should be evaluated

lazappi commented on vignettes/usage_seurat.Rmd:
````diff
@@ -58,6 +58,27 @@ seurat_obj <- adata$as_Seurat()
+adata <- read_zarr(zarr_path)
+seurat_obj_zarr <- adata$as_Seurat()
+seurat_obj_zarr
+```
````
Evaluate examples

lazappi commented on vignettes/usage_singlecellexperiment.Rmd:
````diff
@@ -61,6 +61,27 @@ sce <- adata$as_SingleCellExperiment()
+adata <- read_zarr(zarr_path)
+sce_zarr <- adata$as_SingleCellExperiment()
+sce_zarr
+```
````
Evaluate examples

lazappi reviewed (commented):
A lot of these are TODOs for me 😸

lazappi committed:
Minor text fixes

lazappi committed:
Document .get_compressor

lazappi committed:
Minor fixes to function docs

lazappi committed:
Comment logic in create_zarr_group()

lazappi committed:
Fix indentation in read_zarr_sparse_array()

lazappi committed:
Add construct sparse matrix helper

lazappi committed:
Add ZARR_METADATA_FILES vector

lazappi committed:
Eval Zarr chunks in vignettes

lazappi committed:
Merge test-Zarrv3-read.R into test-Zarr-read.R

lazappi committed:
Combine roundtrip tests

lazappi committed:
Add roundtrip test helpers

lazappi committed:
Refactor test-h5ad-zarr.R to use helper

lazappi committed:
Refactor example files script

- Use uv comments for dependencies
- Update dependencies to latest stable versions
- Add progress messages
- Use Ruff for formatting

lazappi committed:
Remove H5_ITER_INC_ORDERING()

lazappi committed:
Remove Zarr compression comment

lazappi committed:
Fix factor creation in read_zarr_categorical()

lazappi committed:
Pin Rarr version

lazappi committed:
Delete existing Zarr path before writing

lazappi committed:
Add helper functions for accessing Zarr keys

lazappi committed:
Update read_zarr_element() error message

lazappi committed:
Add dimname warnings to ZarrAnnData

lazappi committed:
Add Zarr writeability checks/tests

lazappi committed:
Roxygenise, lint, style

lazappi committed:
Use setup-bioc for all GHA

lazappi committed:
Update WORDLIST

lazappi committed:
Add .venv to .Rbuildignore

lazappi committed:
Clean up test output

lazappi committed:
Style

Artur-man commented:
@lazappi I think I did not check before finishing the hackathon. Do you set the default zarr version for write methods now ? or do you write only zarr v2 or v3 ? 

Again here is the parameter from `anndata` to set the default behaviour:
https://github.com/keller-mark/anndataR/blob/528eff0024f73b39c5147d5eb15e7ad8b9b7bade/inst/scripts/example_files.py#L142

would be nice to do that too if we are not rushing bioc release.

Artur-man commented:
Ok I took another look at it. The roundtrip tests and write tests do not have zarr v3 tests but thats fine for now I think. What is more important is that we can read both versions. Since this is gonna stay as the devel branch for another 6 months I can throw another PR later. Lets not overdo it. 

**UPDATE**: I wrote a version of `test-Zarr-write.R` loops over both v2 and v3 ... looks like its working.

lazappi commented:
I thought there was something that was missing from **{Rarr}** that was blocking writing v3 but I forgot to open an issue and come back to it. If you think the extra tests are passing do you want to push them here?

lazappi commented:
@keller-mark @Artur-man @Bisaloo 

Thanks for your contributions to this PR so far! To properly recognise them feel free to add yourselves as contributors to the `DESCRIPTION`. Or if it's easier comment with your details and I will add you.

Artur-man committed:
add Artur as ctb

rcannood committed:
Merge remote-tracking branch 'origin/devel' into keller-mark/zarr

rcannood committed:
fix linting issues

Bisaloo commented on DESCRIPTION:
```diff
@@ -18,6 +18,8 @@ Authors@R: c(
     person("Chananchida", "Sang-aram", role = "ctb",
            comment = c(ORCID = "0000-0002-0922-0822", github = "csangara")),
+    person("Artür", "Manukyan", role = "ctb",
+           comment = c(ORCID = "0000-0002-0441-9517", github = "Artur-man")),
```
```suggestion
           comment = c(ORCID = "0000-0002-0441-9517", github = "Artur-man")),
    person("Hugo", "Gruson", role = "ctb",
           comment = c(ORCID = "0000-0002-4094-1476", github = "Bisaloo")),
```

Thanks for the offer!

lazappi committed:
Add Bisaloo as contributor

Co-authored-by: Hugo Gruson <hugo.gruson@protonmail.com>

rcannood reviewed (approved):
LGTM!

rcannood commented:
Looks good, thank you all!

The CI is failing, but that's because of the Rarr minimal version requirement. When Bioconductor 3.23 is released, this issue should be fixed.

Artur-man commented:
We did it!!!  1.5 years ... thanks for everyone for the support @keller-mark @Bisaloo @rcannood @lazappi @LouiseDck

ilan-gold commented:
Congrats all!
